src/lib/zork/index.ts

```diff
@@ -234,48 +244,100 @@ const initSocksServer = (ctx: Context) => {
   });
 };
 
-const onDataChannel = (ctx: Context, channel: any) => {
+const handleDataChannel = (ctx: Context, channel: any) => {
   const channelId = channel.label;
-  ctx.log(`onDataChannel: [channel ${channelId}] [channel.readyState ${channel.readyState}]`);
+  ctx.log(`handleDataChannel: [channel ${channelId}] [channel.readyState ${channel.readyState}]`);
 
+  channel.onopen = () => ctx.log(`[channel ${channelId}] datachannel opened`);
   channel.onclose = () => ctx.log(`[channel ${channelId}] datachannel closed`);
   channel.onerror = (err: any) => ctx.log.error(`[channel ${channelId}] datachannel error: ${err}`);
 
-  // ondatachannel events don't fire for the peer creating the datachannel.
-  // Since the getter is always the one that creates datachannels, a getter
-  // doesn't expect any giver-created datachannels, so just closes any onopen.
   if (ctx.mode === 'get') {
-    channel.onopen = () => {
-      channel.close();
-      ctx.log.error(`closed unexpected giver-created datachannel`);
-    };
+    handleGiverCreatedDataChannel(ctx, channel);
     return;
   }
   // Got here -> we're the giver.
 
-  if (channelId === HEARTBEAT_CHANNEL_ID) {  // heartbeat datachannel
-    ctx.socket.end();
-    ctx.log(`closed zork connection, handoff to RTC complete`);
-    numGetters++;
-    ctx.log(`incremented numGetters to ${numGetters}`);
-    ctx.heartbeatTimeoutId = setTimeout(ctx.onHeartbeatTimeout, HEARTBEAT_TIMEOUT_MS);
-    channel.onmessage = (event: any) => {
-      ctx.log(`heartbeat message: "${event.data}"`);
-      clearTimeout(ctx.heartbeatTimeoutId);
-      ctx.heartbeatTimeoutId = setTimeout(ctx.onHeartbeatTimeout, HEARTBEAT_TIMEOUT_MS);
-    };
+  if (channelId === HEARTBEAT_CHANNEL_ID) {
+    handleHeartbeatDataChannel(ctx, channel);
     return;
   }
 
   // Got here -> we're the giver, and this is a non-heartbeat datachannel.
-  // Initialize a SOCKS session for it, and set its onmessage handler so that
-  // it forwards data for the SOCKS session.
-  channel.onopen = () => ctx.log(`[channel ${channelId}] datachannel opened`);
+  handleProxyDataChannel(ctx, channel);
+};
+
+const handleProxyDataChannel = (ctx: Context, channel: any) => {
+  // Initialize a SOCKS session for this datachannel, and set its onmessage
+  // handler so that it forwards data for the SOCKS session.
   initSocksSessionAndFwdSocket(ctx, channel);
-  channel.onmessage = (event: any) => ctx.socksSession.handleDataFromSocksClient(event.data);
+
+  channel.onmessage = (event: any) => {
+    let bytesForSession = event.data;
+    const legacyMessage = typeof event.data === 'string';  // only legacy clients send string messages.
+    if (legacyMessage) {
+      bytesForSession = handleLegacyMessage(ctx, channel, event.data);
+      if (!bytesForSession) {
+        return;
+      }
+    }
+    const session = ctx.socksSessionByChannelId[channel.label];
+    session.handleDataFromSocksClient(bytesForSession);
+  };
+};
```

Comment:
Here we retrieve the right socks session for the given channel using the `ctx.socksSessionByChannelId` map.